src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -113,127 +135,378 @@ internal abstract class AbstractAddParameterCodeFixProvider<
+                                                        .WithAdditionalAnnotations(Formatter.Annotation);
+                    if (anySymbolReferencesNotInSource && methodDeclaration == method)
+                    {
+                        parameterDeclaration = parameterDeclaration.WithAdditionalAnnotations(
+                            ConflictAnnotation.Create(FeaturesResources.Related_method_signatures_found_in_metadata_will_not_be_updated));
+                    }
 
-            var newRoot = editor.GetChangedRoot();
-            var newDocument = methodDocument.WithSyntaxRoot(newRoot);
 
-            return newDocument;
+                    if (method.MethodKind == MethodKind.ReducedExtension)
+                    {
+                        insertionIndex++;
+                    }
+
+                    AddParameter(
+                        syntaxFacts, editor, methodNode, argument,
+                        insertionIndex, parameterDeclaration, cancellationToken);
+                }
+
+                var newRoot = editor.GetChangedRoot();
+                solution = solution.WithDocumentSyntaxRoot(document.Id, newRoot);
+            }
+
+            return solution;
         }
 
-        private async Task<(IParameterSymbol, bool isNamedArgument)> CreateParameterSymbolAsync(
-            Document invocationDocument,
-            IMethodSymbol method,
-            TArgumentSyntax argument,
-            CancellationToken cancellationToken)
+        private static async Task<(ITypeSymbol, RefKind)> GetArgumentTypeAndRefKindAsync(Document invocationDocument, TArgumentSyntax argument, CancellationToken cancellationToken)
         {
             var syntaxFacts = invocationDocument.GetLanguageService<ISyntaxFactsService>();
-            var semanticFacts = invocationDocument.GetLanguageService<ISemanticFactsService>();
-            var argumentName = syntaxFacts.GetNameForArgument(argument);
-            var expression = syntaxFacts.GetExpressionOfArgument(argument);
-
             var semanticModel = await invocationDocument.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-            var parameterType = semanticModel.GetTypeInfo(expression).Type ?? semanticModel.Compilation.ObjectType;
+            var argumentExpression = syntaxFacts.GetExpressionOfArgument(argument);
+            var argumentType = semanticModel.GetTypeInfo(argumentExpression).Type ?? semanticModel.Compilation.ObjectType;
+            var refKind = syntaxFacts.GetRefKindOfArgument(argument);
+            return (argumentType, refKind);
+        }
+
+        private static async Task<ImmutableArray<IMethodSymbol>> FindMethodDeclarationReferences(
+            Document invocationDocument, IMethodSymbol method, CancellationToken cancellationToken)
+        {
+            var progress = new StreamingProgressCollector(StreamingFindReferencesProgress.Instance);
```

Comment:
@dpoeschl I believe your signature change service - http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/ChangeSignature/AbstractChangeSignatureService.cs,a6348ee4b0207df5

handle most of this cases? except currently it doesn't support adding parameter. is that hard to add support there? adding parameter? considering it already has most of functionality this fix is trying to do? like changing all definitions, references, delegates and etc?